File: week_02/java/src/main/java/com/example/NER_and_Stopwords.java
package com.example;

import kr.co.shineware.nlp.komoran.constant.DEFAULT_MODEL;
import kr.co.shineware.nlp.komoran.core.Komoran;
import kr.co.shineware.nlp.komoran.model.KomoranResult;
import kr.co.shineware.nlp.komoran.model.Token;

import java.util.*;

public class NER_and_Stopwords {

    Map<String, String> NER_MAP;
    Set<String> STOP_WORDS;

    public void init() {

        NER_MAP = new HashMap<>();
        NER_MAP.put("서울", "LOC");
        NER_MAP.put("데이터분석", "SKL");

        STOP_WORDS = new HashSet<>();
        STOP_WORDS.add("안녕하세요");

    }

    public void action() {

        this.init();

        Komoran komoran = new Komoran(DEFAULT_MODEL.FULL);
        String document = "안녕하세요 저는 서울에 사는 철수입니다. 데이터분석은 참 재미있는 것 같아요.";

        KomoranResult analyzeResultList = komoran.analyze(document);
        System.out.println(analyzeResultList.getPlainText());

        List<Token> tokenList = analyzeResultList.getTokenList();
        for (Token token : tokenList) {
            System.out.format("(%2d, %2d) %s/%s\n", token.getBeginIndex(),
                    token.getEndIndex(), token.getMorph(), token.getPos());
        }

        List<String> nounList = getNounList(analyzeResultList.getNouns());
        System.out.println("명사 리스트 -> " + nounList);
        // Map<String, String> nerMap = getNer
        Map<String, String> nerMap = getNerMap(tokenList);
        System.out.println("NER Map -> " + nerMap);

    }

    public List<String> getNounList(List<String> nounList) {

        List<String> rtnList = new ArrayList<>();
        for (String noun : nounList) {
            if (STOP_WORDS.contains(noun))
                continue;
            rtnList.add(noun);
        }

        return rtnList;

    }

    public Map<String, String> getNerMap(List<Token> tokenList) {

        Map<String, String> nerMap = new HashMap<>();
        String beforeStr = "";
        // boolean flags = false;

        for (Token token : tokenList) {
            // 명사가 아닐 경우, 초기화
            if (token.getPos().indexOf("NN") == -1) {
                beforeStr = "";
                continue;
            }

            // 명사일 경우
            String mergeStr = beforeStr + token.getMorph();
            if (NER_MAP.containsKey(mergeStr)) {
                nerMap.put(mergeStr, NER_MAP.get(mergeStr));
            }

            beforeStr = token.getMorph();
        }

        return nerMap;

    }

}

File: week_02/python/ner_and_stopwords.py
from konlpy.tag import Komoran
from konlpy.utils import pprint


class NER_and_Stopwords:
    
    def __init__(self):
        
        self.NER_MAP = dict()
        self.NER_MAP["서울"] = "LOC"
        self.NER_MAP["데이터분석"] = "SKL"
        
        self.STOPWORDS = set()
        self.STOPWORDS.add("안녕하세요")
        self.STOPWORDS.add("것")
    
    def action(self):
        
        komoran = Komoran()
        document = "안녕하세요. 저는 서울에 사는 철수입니다. 데이터분석은 참 재미있는 것 같아요."
        
        tokenPosList = komoran.pos(document)
        pprint(tokenPosList)
        
        nounList = self.get_nouns(komoran.nouns(document))
        print("NOUN LIST -> ", nounList)
        
        nerMap = self.get_ner_map(tokenPosList)
        print("NER MAP -> ", nerMap)
        
        return None
    
    def get_nouns(self, nounList: list):
        
        rtnList = []
        for n in nounList:
            if n in self.STOPWORDS:
                continue
            rtnList.append(n)
        
        return rtnList
    
    def get_ner_map(self, tokenPosList):
        
        nerMap = dict()
        beforeStr = ""
        for token, pos in tokenPosList:
            if "NN" not in pos: # 명사가 아닐 경우
                beforeStr = ""
                continue
            
            mergeStr = beforeStr + token
            if mergeStr in self.NER_MAP.keys():
                nerMap[mergeStr] = self.NER_MAP[mergeStr]
            
            beforeStr = token
        
        return nerMap
    